src/docfx/restore/RestoreMap.cs

```diff
@@ -9,40 +9,47 @@
 
 namespace Microsoft.Docs.Build
 {
-    internal class RestoreMap
+    internal static class RestoreMap
     {
-        private static readonly ConcurrentDictionary<string, Lazy<string>> s_gitPath = new ConcurrentDictionary<string, Lazy<string>>();
+        private static readonly ConcurrentDictionary<(string remote, string branch), Lazy<string>> s_gitPath = new ConcurrentDictionary<(string remote, string branch), Lazy<string>>();
         private static readonly ConcurrentDictionary<string, Lazy<string>> s_downloadPath = new ConcurrentDictionary<string, Lazy<string>>();
 
-        private readonly string _docsetPath;
-
-        public RestoreMap(string docsetPath)
+        public static string GetGitRestorePath(string url)
         {
-            Debug.Assert(!string.IsNullOrEmpty(docsetPath));
-            _docsetPath = docsetPath;
+            var (remote, branch) = GitUtility.GetGitRemoteInfo(url);
+            return GetGitRestorePath(remote, branch);
         }
 
-        public string GetGitRestorePath(string url)
+        public static string GetGitRestorePath(string remote, string branch)
         {
-            Debug.Assert(!string.IsNullOrEmpty(url));
+            if (!TryGetGitRestorePath(remote, branch, out var result))
+            {
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+            return result;
+        }
 
-            var gitRestorePath = s_gitPath.GetOrAdd(url, new Lazy<string>(FindLastModifiedGitRepository)).Value;
+        public static bool TryGetGitRestorePath(string remote, string branch, out string result)
+        {
+            result = s_gitPath.GetOrAdd((remote, branch), new Lazy<string>(FindLastModifiedGitRepository)).Value;
 
-            if (!Directory.Exists(gitRestorePath))
+            // GetOrAdd and TryRemove together are not atomic operation,
+            // but because we only call TryGetGitRestorePath in implicit restore,
+            // it is very unlikely to cause a problem.
+            if (result == null)
             {
-                throw Errors.NeedRestore(url).ToException();
+                s_gitPath.TryRemove((remote, branch), out _);
```

Comment:
Why do we need to remove the null value? Since the return result is false, the required key `remote, branch` is not found.

Reply:
`s_gitPath` will cache `null`. If a folder does not exist and is later restored, the cache is still `null`.